### ISSUE TITLE
Javadoc: Incorrect parsing of white space token as text for link literal

### ISSUE THREAD

**Issue by AkMo3** [checkstyle/checkstyle#10028]
I have read check documentation: https://checkstyle.org/config_xxxxxx.html#NameOfAffectedCheck
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
$ javac Test1.java
                                     
$ cat Test1.java 
public class Test1 {
       
  /**
   * {@link www.google.com}
   */
   void foo1() {}
}

$ java -jar checkstyle-8.43-SNAPSHOT-all.jar -J Test1.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test1 [1:13]
`--OBJBLOCK -> OBJBLOCK [1:19]
    |--LCURLY -> { [1:19]
    |--METHOD_DEF -> METHOD_DEF [6:3]
    |   |--MODIFIERS -> MODIFIERS [6:3]
    |   |--TYPE -> TYPE [6:3]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [3:2]
    |   |   |   |--COMMENT_CONTENT -> *\n   * {@link www.google.com}\n    [3:4]
    |   |   |   |   `--JAVADOC -> JAVADOC [3:5]
    |   |   |   |       |--NEWLINE -> \n [3:5]
    |   |   |   |       |--LEADING_ASTERISK ->    * [4:0]
    |   |   |   |       |--TEXT ->   [4:4]                                 // Error,  Expected WS-> [4:4]
    |   |   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [4:5]
    |   |   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [4:5]
    |   |   |   |       |   |--LINK_LITERAL -> @link [4:6]
    |   |   |   |       |   |--WS ->   [4:11]
    |   |   |   |       |   |--REFERENCE -> REFERENCE [4:12]
    |   |   |   |       |   |   `--PACKAGE_CLASS -> www.google.com [4:12]
    |   |   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [4:26]
    |   |   |   |       |--NEWLINE -> \n [4:27]
    |   |   |   |       |--TEXT ->     [5:0]
    |   |   |   |       `--EOF -> <EOF> [5:3]
    |   |   |   `--BLOCK_COMMENT_END -> */ [5:3]
    |   |   `--LITERAL_VOID -> void [6:3]
    |   |--IDENT -> foo1 [6:8]
    |   |--LPAREN -> ( [6:12]
    |   |--PARAMETERS -> PARAMETERS [6:13]
    |   |--RPAREN -> ) [6:13]
    |   `--SLIST -> { [6:15]
    |       `--RCURLY -> } [6:16]
    `--RCURLY -> } [7:0]
```
---

Expected White Space Token

This parsing is done correctly for others. Eg:

For 
```
  /**
   * @code Something.
   */
   void foo1() {}
```

AST: 
```
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [3:2]
    |   |   |   |--COMMENT_CONTENT -> *\n   * @code Something.\n    [3:4]
    |   |   |   |   `--JAVADOC -> JAVADOC [3:5]
    |   |   |   |       |--NEWLINE -> \n [3:5]
    |   |   |   |       |--LEADING_ASTERISK ->    * [4:0]
    |   |   |   |       |--WS ->   [4:4]
    |   |   |   |       |--JAVADOC_TAG -> JAVADOC_TAG [4:5]
    |   |   |   |       |   |--CUSTOM_NAME -> @code [4:5]
    |   |   |   |       |   |--WS ->   [4:10]
    |   |   |   |       |   `--DESCRIPTION -> DESCRIPTION [4:11]
    |   |   |   |       |       |--TEXT -> Something. [4:11]
    |   |   |   |       |       |--NEWLINE -> \n [4:21]
    |   |   |   |       |       `--TEXT ->     [5:0]
    |   |   |   |       `--EOF -> <EOF> [5:3]
    |   |   |   `--BLOCK_COMMENT_END -> */ [5:3]
```

**Comment by AkMo3** [comment]
@nmancus1 Please share your view

**Comment by nmancus1** [comment]
This will happen for any alternative in our start rule (`javadoc`) or some other rules (like `description`):

```antlr

javadoc: (
            htmlElement
            | ({!isNextJavadocTag()}? LEADING_ASTERISK)
            | htmlComment
            | NEWLINE
            | text
            | javadocInlineTag
         )*                                      // first closure
         (LEADING_ASTERISK? WS* javadocTag)*
         EOF;
```

This is because our `text` rule captures WS:

```antlr

text : ((CHAR | WS)
 {
  _la = _input.LA(1);
  if ((_la != WS) && (_la != CHAR)) return _localctx;
  else if (_alt == 1) continue;
 }
       )+;

```

Any whitespace encountered upon parsing of the alternatives in the first closure in the `javadoc`, `description`, and similar rules, matches the `text` rule. 

I am not sure if this was by design or not, but, changing this would mean modifying our start rule and other heavily used productions; this means that we will have a lot of tests to fix, and probably some broken checks. I am not sure if it worth it to change this, after all, whitespace is text. But, it would make our javadoc AST more consistent. There are a couple of ways that this could be achieved; one is to simply add a `WS` or `WS+` alternative to the first close as mentioned above.

**Comment by AkMo3** [comment]
I will conduct a feasibility study about how much the change will be. Currently, I see 43 test failure. It would take some time to fix everything.

**Comment by AkMo3** [comment]
@nmancus1

I see this issue in html tags also. Like in 
```
class TestClass {
    /**
     * <p> This is Summary</p>
     */
    public void method(int param1, int param2) {
    }
}
```
 
The ast of our focus is:
```
    |   |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [3:7]
    |   |   |   |       |   `--PARAGRAPH -> PARAGRAPH [3:7]
    |   |   |   |       |       |--P_TAG_START -> P_TAG_START [3:7]
    |   |   |   |       |       |   |--START -> < [3:7]
    |   |   |   |       |       |   |--P_HTML_TAG_NAME -> p [3:8]
    |   |   |   |       |       |   `--END -> > [3:9]
    |   |   |   |       |       |--TEXT ->  This is Summary [3:10]
    |   |   |   |       |       `--P_TAG_END -> P_TAG_END [3:26]
    |   |   |   |       |           |--START -> < [3:26]
    |   |   |   |       |           |--SLASH -> / [3:27]
    |   |   |   |       |           |--P_HTML_TAG_NAME -> p [3:28]
    |   |   |   |       |           `--END -> > [3:29]
```

A better ast can be:
```
    |   |   |   |       |--HTML_ELEMENT -> HTML_ELEMENT [3:7]
    |   |   |   |       |   `--PARAGRAPH -> PARAGRAPH [3:7]
    |   |   |   |       |       |--P_TAG_START -> P_TAG_START [3:7]
    |   |   |   |       |       |   |--START -> < [3:7]
    |   |   |   |       |       |   |--P_HTML_TAG_NAME -> p [3:8]
    |   |   |   |       |       |   `--END -> > [3:9]
    |   |   |   |       |       |--WS ->   [3:10]
    |   |   |   |       |       |--TEXT -> This is Summary [3:11]
    |   |   |   |       |       `--P_TAG_END -> P_TAG_END [3:26]
    |   |   |   |       |           |--START -> < [3:26]
    |   |   |   |       |           |--SLASH -> / [3:27]
    |   |   |   |       |           |--P_HTML_TAG_NAME -> p [3:28]
    |   |   |   |       |           `--END -> > [3:29]
```

Should I include this issue also along with. I tring to see how much changes is required, and if they are feasibile also becomes what seems from now is it is going to be big change. But since you would soon be starting your work on new ANTLR, your view is important about the the design.

**Comment by nmancus1** [comment]
> I see this issue in html tags...

Correct. As stated above, this will happen with any subrules in the first closure.

After thinking on this a bit, I am not sure if there really is a problem here. These inline tags are part of the initial description of the Javadoc. So, if there is any element in a javadoc that precedes a block tag, it is part of the description. I cannot find a Javavdoc grammar from Oracle (please share if you know where to find one), but in the [Javadoc documentation](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#format), it states that:

> A doc comment is written in HTML and must precede a class, field, constructor or method declaration. It is made up of two parts -- a description followed by block tags.

To me, this sounds like:

```
S -> description blockTags EOF
...
```

Which  is basically what we have:

```
javadoc: (
            htmlElement
            | ({!isNextJavadocTag()}? LEADING_ASTERISK)
            | htmlComment
            | NEWLINE
            | text
            | javadocInlineTag
         )*                                      // description
         (LEADING_ASTERISK? WS* javadocTag)* // blockTags
         EOF; // eof

```

Then, also from the Javadoc documentation:

>Notice the inline tag {@link URL}, which converts to an HTML hyperlink pointing to the documentation for the URL class. This inline tag can be used anywhere that a comment can be written, such as in the text following block tags.

also from https://doc.bccnsoft.com/docs/jdk7-docs/technotes/tools/solaris/javadoc.html:

> Can be placed anywhere in the main description or in the comments for block tags. Inline tags are denoted by curly braces: {@tag}.

So inline tags are  productions that could appear **within** block tags and the main description. Inline tags are a lesser element. So, this:

```java

/**
 * My description {@code x}.
 */
public class Test {
    
}

```

and this:

```java

/**
 *   {@code x}
 */
public class Test {
    
}

```

and this:

```java

/**
 *  My description.
 */
public class Test {
    
}

```

are Javadocs comprised of only a `description` ,  and all of the elements they contain are children/ sub-elements of the main description. We should not treat a space before an inline tag (`  {@code x}`) differently than a space in the body of a main description (`My description.`), because they are lexically the same thing.

**Comment by nmancus1** [comment]
> A better ast can be...

Please explain why this is better.

**Comment by AkMo3** [comment]
> We should not treat whitespace before an inline tag ( {@code x}) differently than whitespace in the body of a main description (My description.), because they are lexically the same thing.

I am working on this only, such that all whitespaces before text are classified as `TokenTypes.WS`

> Please explain why this is better.

1. I believe it is better because in certain cases after getting the text of an element, we are not sure if it empty of not, so we are required to use `JavadocUtil.isEmpty()`

2. For cases like

    ```
    /**
     *          This is Summary.
     */
    public void method(int param1, int param2) {
    }
    ```
    The produced AST for Javadoc is
    ```
    JAVADOC -> JAVADOC [2:7]
     |--NEWLINE -> \r\n [2:7]
     |--LEADING_ASTERISK ->      * [3:0]
     |--WS ->            [3:6]
     |--TEXT -> This is Summary. [3:16]
     |--NEWLINE -> \r\n [3:32]
     |--WS ->       [4:0]
      `--EOF -> <EOF> [4:5]
     ```
     
    This implementation will come in handy for us while enforcing that there are no white or single white spaces in certain conditions.
          
> I cannot find a Javadoc grammar from Oracle (please share if you know where to find one)

I could not find one that matches our need. The javadoc grammar isn't published by oracle. And I guess the checkstyle javadoc grammar is the most comprehensive one. Rest are very basic implementation of javadoc with no lexers of inline tag and html characters etc.

**Comment by nmancus1** [comment]
> such that all whitespaces before text are classified as `TokenTypes.WS`

Another argument for this is that we should show the leading whitespace separately, because the javadoc tool trims it anyway:

```bash
➜  src cat Test.java
/**
 *                My description.
 */
public class Test {}


```
Renders as (edge of window shown for reference):

![image](https://user-images.githubusercontent.com/31252532/119492615-4690ba00-bd2d-11eb-9ff5-628145e36b3b.png)

But then, it does this for all lines:

```bash

➜  src cat Test.java
/**
 *                  Some description
 *                         {@link www.github.com}
 *                                 other description
 */
public class Test {}

```
Renders as:
![image](https://user-images.githubusercontent.com/31252532/119493833-9d4ac380-bd2e-11eb-9d7a-d6703a63b8be.png)

So, should all preceding whitespace on each line become `WS`?

**Comment by pbludov** [comment]
Maybe this space should be part of the leading asterisk.

**Comment by AkMo3** [comment]
> So, should all preceding whitespace on each line become WS? Or, only the first line?

I am making a draft PR for all changes for all changes where all whitespaces before text are classified as `WS`. 

Eg: 

File: 
```java
/**
 *                  Some description
 *                         {@link www.github.com}
 *                                 other description
 */
public class Test {}
```

AST: 
```AST
CLASS_DEF -> CLASS_DEF [6:0]
|--MODIFIERS -> MODIFIERS [6:0]
|   |--BLOCK_COMMENT_BEGIN -> /* [1:0]
|   |   |--COMMENT_CONTENT -> *\r\n *                  Some description\r\n *                         {@link www.github.com}\r\n *                                 other description\r\n  [1:2]
|   |   |   `--JAVADOC -> JAVADOC [1:3]
|   |   |       |--NEWLINE -> \r\n [1:3]
|   |   |       |--LEADING_ASTERISK ->  * [2:0]
|   |   |       |--WS ->                    [2:2]
|   |   |       |--TEXT -> Some description [2:20]
|   |   |       |--NEWLINE -> \r\n [2:36]
|   |   |       |--LEADING_ASTERISK ->  * [3:0]
|   |   |       |--WS ->                           [3:2]
|   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [3:27]
|   |   |       |   |--JAVADOC_INLINE_TAG_START -> { [3:27]
|   |   |       |   |--LINK_LITERAL -> @link [3:28]
|   |   |       |   |--WS ->   [3:33]
|   |   |       |   |--REFERENCE -> REFERENCE [3:34]
|   |   |       |   |   `--PACKAGE_CLASS -> www.github.com [3:34]
|   |   |       |   `--JAVADOC_INLINE_TAG_END -> } [3:48]
|   |   |       |--NEWLINE -> \r\n [3:49]
|   |   |       |--LEADING_ASTERISK ->  * [4:0]
|   |   |       |--WS ->                                   [4:2]
|   |   |       |--TEXT -> other description [4:35]
|   |   |       |--NEWLINE -> \r\n [4:52]
|   |   |       |--WS ->   [5:0]
|   |   |       `--EOF -> <EOF> [5:1]
|   |   `--BLOCK_COMMENT_END -> */ [5:1]
|   `--LITERAL_PUBLIC -> public [6:0]
|--LITERAL_CLASS -> class [6:7]
|--IDENT -> Test [6:13]
`--OBJBLOCK -> OBJBLOCK [6:18]
    |--LCURLY -> { [6:18]
    `--RCURLY -> } [6:19]
```

**Comment by gianmarcoschifone** [comment]
@romani I rechecked this on current upstream master.

The behavior is still reproducible: whitespace before the inline tag is emitted as `TEXT`, not `WS`.

```text
$ cat Test.java
public class Test {

  /**
   * {@link www.google.com}
   */
   void foo1() {}
}

$ java -jar target/checkstyle-13.8.0-SNAPSHOT-all.jar -J Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:1]
`--CLASS_DEF -> CLASS_DEF [1:1]
    |--MODIFIERS -> MODIFIERS [1:1]
    |   `--LITERAL_PUBLIC -> public [1:1]
    |--LITERAL_CLASS -> class [1:8]
    |--IDENT -> Test [1:14]
    `--OBJBLOCK -> OBJBLOCK [1:19]
        |--LCURLY -> { [1:19]
        |--METHOD_DEF -> METHOD_DEF [6:4]
        |   |--MODIFIERS -> MODIFIERS [6:4]
        |   |--TYPE -> TYPE [6:4]
        |   |   |--BLOCK_COMMENT_BEGIN -> /* [3:3]
        |   |   |   |--COMMENT_CONTENT -> *\n   * {@link www.google.com}\n    [3:5]
        |   |   |   |   `--JAVADOC_CONTENT -> JAVADOC_CONTENT [3:6]
        |   |   |   |       |--NEWLINE -> \n [3:6]
        |   |   |   |       |--LEADING_ASTERISK ->    * [4:1]
        |   |   |   |       |--TEXT ->   [4:5]
        |   |   |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [4:6]
        |   |   |   |       |   `--LINK_INLINE_TAG -> LINK_INLINE_TAG [4:6]
        |   |   |   |       |       |--JAVADOC_INLINE_TAG_START -> {@ [4:6]
        |   |   |   |       |       |--TAG_NAME -> link [4:8]
        |   |   |   |       |       |--TEXT ->   [4:12]
        |   |   |   |       |       |--REFERENCE -> REFERENCE [4:13]
        |   |   |   |       |       |   `--IDENTIFIER -> www.google.com [4:13]
        |   |   |   |       |       `--JAVADOC_INLINE_TAG_END -> } [4:27]
        |   |   |   |       |--NEWLINE -> \n [4:28]
        |   |   |   |       `--TEXT ->     [5:1]
        |   |   |   `--BLOCK_COMMENT_END -> */ [5:4]
        |   |   `--LITERAL_VOID -> void [6:4]
        |   |--IDENT -> foo1 [6:9]
        |   |--LPAREN -> ( [6:13]
        |   |--PARAMETERS -> PARAMETERS [6:14]
        |   |--RPAREN -> ) [6:14]
        |   `--SLIST -> { [6:16]
        |       `--RCURLY -> } [6:17]
        `--RCURLY -> } [7:1]
```

This is only an internal AST-shape inconsistency. In my opinion, it does not need to be fixed and should remain low priority unless checks need this whitespace normalized.

**Comment by romani** [comment]
lets not fix this, no much interest to this AST from community, we can change our opinion when we get more requests from community.
